routing/graph.go

```diff
@@ -5,110 +5,54 @@ import (
-	sourceNode() route.Vertex
-
-	// fetchNodeFeatures returns the features of the given node.
-	fetchNodeFeatures(nodePub route.Vertex) (*lnwire.FeatureVector, error)
-
-	// FetchAmountPairCapacity determines the maximal capacity between two
-	// pairs of nodes.
-	FetchAmountPairCapacity(nodeFrom, nodeTo route.Vertex,
-		amount lnwire.MilliSatoshi) (btcutil.Amount, error)
-}
-
-// CachedGraph is a routingGraph implementation that retrieves from the
-// database.
-type CachedGraph struct {
-	graph  *channeldb.ChannelGraph
-	tx     kvdb.RTx
-	source route.Vertex
-}
-
-// A compile time assertion to make sure CachedGraph implements the routingGraph
-// interface.
-var _ routingGraph = (*CachedGraph)(nil)
-
-// NewCachedGraph instantiates a new db-connected routing graph. It implicitly
-// instantiates a new read transaction.
-func NewCachedGraph(sourceNode *channeldb.LightningNode,
-	graph *channeldb.ChannelGraph) (*CachedGraph, error) {
-
-	tx, err := graph.NewPathFindTx()
-	if err != nil {
-		return nil, err
-	}
-
-	return &CachedGraph{
-		graph:  graph,
-		tx:     tx,
-		source: sourceNode.PubKeyBytes,
-	}, nil
+	// FetchNodeFeatures returns the features of the given node.
+	FetchNodeFeatures(nodePub route.Vertex) (*lnwire.FeatureVector, error)
 }
 
-// Close attempts to close the underlying db transaction. This is a no-op in
-// case the underlying graph uses an in-memory cache.
-func (g *CachedGraph) Close() error {
-	if g.tx == nil {
-		return nil
-	}
+// GraphSession describes a new read session with a Graph backend. It must be
+// closed using its Close method after path-finding is complete.
+type GraphSession interface {
+	Graph
 
-	return g.tx.Rollback()
+	// Close closes the GraphSession and must be called once path-finding
+	// using this GraphSession is complete.
+	Close() error
 }
 
-// forEachNodeChannel calls the callback for every channel of the given node.
-//
-// NOTE: Part of the routingGraph interface.
-func (g *CachedGraph) forEachNodeChannel(nodePub route.Vertex,
-	cb func(channel *channeldb.DirectedChannel) error) error {
-
-	return g.graph.ForEachNodeDirectedChannel(g.tx, nodePub, cb)
-}
-
-// sourceNode returns the source node of the graph.
-//
-// NOTE: Part of the routingGraph interface.
-func (g *CachedGraph) sourceNode() route.Vertex {
-	return g.source
-}
-
-// fetchNodeFeatures returns the features of the given node. If the node is
-// unknown, assume no additional features are supported.
-//
-// NOTE: Part of the routingGraph interface.
-func (g *CachedGraph) fetchNodeFeatures(nodePub route.Vertex) (
-	*lnwire.FeatureVector, error) {
-
-	return g.graph.FetchNodeFeatures(nodePub)
+// GraphSessionFactory can be used to produce a new GraphSession instance which
+// can then be used for a path-finding session. Depending on the implementation,
+// the GraphSession will represent a DB connection where a read-lock is being
+// held across calls to the backing Graph.
+type GraphSessionFactory interface {
```

Comment:
the whole thing is so that the logic can remain as is today where the payment session creates a ReadTx to perform a couple of calls over to ensure db consistency over those calls.

Reply:
note that this read lock is anyways not used if the `graphCach` is being used. So potentially we can just change PaymentSession to not require this lock anyways... but yeah - again just trying to keep this PR to be a pure refactor. 

but if we all agree that it isnt that important for the lock to be held throughout the payment session, then I can add a commit at the start of this PR to change that. And then yes that would make this part way simpler.

Reply:
> ReadTx to perform a couple of calls over to ensure db consistency over those calls

Can we try to hide the `kvdb.RwTx` from the implementor of the interface? Given all we need is the ability to create that txn, then "close" it. If someone is passing in say a purely in memory graph loaded from RPC, then there's no real txn concept, but we can maintain it in a slightly more abstracted manner.